### DISCUSSION TITLE
Creating a base class with customized table names for all subclasses

### DISCUSSION THREAD

**Discussion by davetron5000** [jeremyevans/sequel#2216] [General]
I'm trying to make a base class for a `Sequel::Model` that does two things:

* Establishes a common base class where shared functions can be added
* Changes the "algorithm" for how a table name is determined

I know this isn't the right code, but conceptually:

```ruby
class BaseDataModel < Sequel::Model
  def table_name
    self.class.name.gsub(/DataModel$/,"").underscorize.pluarlize
  end
end

class AccountsDataModel < BaseDataModel
end

AccountsDataModel[name: "pat"] # => select * from accounts where name = 'pat'
```

The base class part is no problem:

```ruby
AppDataModel = Class.new(Sequel::Model)
```

I can't figure out how to change the table name "algorithm".  I tried implementing `inherited` and calling `set_dataset` in there, but that caused `undefined method `internal_name=' for an instance of DataModel::EntitlementDefault` which told me that I should not have done what I did.

So, question is - how might I do this?

Let's set aside if this is a good idea or not, just looking to understand how it can be done.

**Comment by jeremyevans** [comment]
You want to override the `implicit_table_name` method.  See the implementation of the singular_table_names plugin.

**Comment by davetron5000** [comment]
Ah, that worked, thank you!
